src/src/com/microsoft/aad/adal/AcquireTokenSilentHandler.java

```diff
@@ -0,0 +1,308 @@
+            // already tried our best, null will be retured. If it eixsts, try with it. 
+            // If we have already tried an MRRT and no FRT found, we return the MRRT result passed in. 
+            if (!mAttemptedWithMRRT) {
+                Logger.v(TAG, "FRT cache item does not exist, fall back to try MRRT.");
+                return useMRRT();
+            } else {
+                return mrrtResult;
+            }
+        }
+        
+        Logger.v(TAG, "Send request to use FRT for new AT.");
+        AuthenticationResult frtResult = acquireTokenWithCachedItem(frtTokenCacheItem);
+        if (isTokenRequestFailed(frtResult) && !mAttemptedWithMRRT) {
+            // FRT request fails, fallback to MRRT if we haven't tried with MRRT. 
+            final AuthenticationResult retryMrrtResult = useMRRT();
+            frtResult = retryMrrtResult == null ? frtResult : retryMrrtResult;
+        }
+        
+        return frtResult;
+    }
+
+    /**
+     * Attempt to use MRRT. 
+     */
+    private AuthenticationResult useMRRT() throws AuthenticationException {
+        Logger.v(TAG, "Send request to use MRRT for new AT.");
+        mAttemptedWithMRRT = true;
+        if (mMrrtTokenCacheItem == null) {
+            return null;
+        }
+        
+        return acquireTokenWithCachedItem(mMrrtTokenCacheItem);
+    }
+    
+    /**
+     * Acquire token with retrieved token cache item and update cache. 
+     */
+    private AuthenticationResult acquireTokenWithCachedItem(final TokenCacheItem cachedItem)
+            throws AuthenticationException {
+        final AuthenticationResult result = acquireTokenWithRefreshToken(cachedItem.getRefreshToken());
+        
+        if (result != null) {
+            mTokenCacheAccessor.updateCachedItemWithResult(mAuthRequest.getResource(), mAuthRequest.getClientId(), 
+                    result, cachedItem);
+        }
+
+        return result;
+    }
+    
+    /**
+     * Old version of ADAL doesn't mark token stored in regular RT entry as MRRT even it is. The logic to look for 
+     * MRRT is when RT is not found or found RT is also MRRT. To support the old behavior, do a separate check on
+     * the existence for MRRT token entry. 
+     */
+    private boolean isMRRTEntryExisted() {
+        final TokenCacheItem mrrtItem = mTokenCacheAccessor.getMRRTItem(mAuthRequest.getClientId(), getUserFromRequest());
+        return mrrtItem != null && !StringExtensions.IsNullOrBlank(mrrtItem.getRefreshToken());
+    }
+    
+    /**
+     * Check if the {@link AuthenticationResult} contains oauth2 error. 
+     */
+    private boolean isTokenRequestFailed(final AuthenticationResult result) {
+        return result != null && !StringExtensions.IsNullOrBlank(result.getErrorCode());
+    }
+    
+    /**
+     * Get either loginhint or user id based what's passed in the request. 
+     */
+    private String getUserFromRequest() {
```

Comment:
Done. 

---
In reply to: [63267579](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/611#discussion_r63267579) [](ancestors = 63267579)